metals/src/main/scala/scala/meta/internal/metals/OutlineFilesProvider.scala

```diff
@@ -0,0 +1,191 @@
+            provider.wasSuccessfullyCompiledByBuildServer
+          case None => true
+        }
+    }
+  }
+
+  def onDidCompile(id: BuildTargetIdentifier, wasSuccessful: Boolean): Unit = {
+    outlineFiles.get(id) match {
+      case Some(provider) =>
+        if (wasSuccessful) provider.successfulCompilation()
+      case None =>
+        for {
+          scalaTarget <- buildTargets.scalaTarget(id)
+          // we don't perform outline compilation for Scala 3
+          if (!ScalaVersions.isScala3Version(scalaTarget.scalaVersion))
+        } outlineFiles.putIfAbsent(
+          id,
+          new BuildTargetOutlineFilesProvider(
+            buildTargets,
+            buffers,
+            id,
+            wasSuccessful,
+          ),
+        )
+    }
+  }
+
+  def didChange(id: BuildTargetIdentifier, path: AbsolutePath): Unit = {
+    for {
+      provider <- outlineFiles.get(id)
+    } provider.didChange(path)
+  }
+
+  def getOutlineFiles(
+      buildTargetId: Option[BuildTargetIdentifier]
+  ): Optional[JOutlineFiles] = {
+    val res: Option[JOutlineFiles] =
+      for {
+        id <- buildTargetId
+        provider <- outlineFiles.get(id)
+        outlineFiles <- provider.outlineFiles()
+      } yield outlineFiles
+    res.asJava
+  }
+
+  def enrichWithOutlineFiles(
+      buildTargetId: Option[BuildTargetIdentifier]
+  )(vFile: CompilerVirtualFileParams): CompilerVirtualFileParams = {
+    val optOutlineFiles =
+      for {
+        id <- buildTargetId
+        provider <- outlineFiles.get(id)
+        outlineFiles <- provider.outlineFiles()
+      } yield outlineFiles
+
+    optOutlineFiles
+      .map(outlineFiles => vFile.copy(outlineFiles = Optional.of(outlineFiles)))
+      .getOrElse(vFile)
+  }
+
+  def enrichWithOutlineFiles(
+      path: AbsolutePath
+  )(vFile: CompilerVirtualFileParams): CompilerVirtualFileParams = {
+    val optOutlineFiles =
+      for {
+        bt <- buildTargets.inferBuildTarget(path)
+        provider <- outlineFiles.get(bt)
+        outlineFiles <- provider.outlineFiles()
+      } yield outlineFiles
+
+    optOutlineFiles
+      .map(outlineFiles => vFile.copy(outlineFiles = Optional.of(outlineFiles)))
+      .getOrElse(vFile)
+  }
+
+  def clear(): Unit = {
+    outlineFiles.clear()
+  }
+}
+
+class BuildTargetOutlineFilesProvider(
+    buildTargets: BuildTargets,
+    buffers: Buffers,
+    id: BuildTargetIdentifier,
+    wasCompilationSuccessful: Boolean,
+) {
+  private val changedDocuments =
+    ConcurrentHashSet.empty[AbsolutePath]
+
+  private val wasAllOutlined: AtomicBoolean =
+    new AtomicBoolean(false)
+
+  private val wasSuccessfullyCompiled: AtomicBoolean =
+    new AtomicBoolean(wasCompilationSuccessful)
+
+  def wasSuccessfullyCompiledByBuildServer: Boolean =
+    wasSuccessfullyCompiled.get()
+
+  def successfulCompilation(): Unit = {
+    wasSuccessfullyCompiled.set(true)
+    changedDocuments.clear()
```

Comment:
Note: that this doesn't account for possible changes before compilation end is reported, it may not 100% reflect the state of files that was during compilation.